compiler/src/iree/compiler/Codegen/Common/GPU/GPUMaterializeEncoding.cpp

```diff
@@ -36,8 +39,10 @@ getIntrinsicVectorSize(TypeRange elementTypes, int64_t roleIdx) {
   Type rhs = elementTypes[1];
   Type out = elementTypes[2];
   if (lhs.isF32() && rhs.isF32() && out.isF32()) {
-    if (roleIdx == 0 || roleIdx == 1) return std::make_pair(1, 1);
-    if (roleIdx == 2) return std::make_pair(4, 1);
+    if (roleIdx == 0 || roleIdx == 1)
+      return std::make_pair(1, 1);
+    if (roleIdx == 2)
+      return std::make_pair(4, 1);
```

Comment:
The vector sizes for an intrinsic will not necessarily always be rank 2. I think this should be a vector instead of a pair.

For example, the C matrix layout of the `v_mfma_f32_32x32x8_f16` intrinsic requires 3 inner vector sizes even though there are only 2 dimensions in the inner tile block. You can see this in the tensor layout here: https://gist.github.com/Max191/145b1ad58c5fdabbd656022f5c00da0b

Lane 0, 1, 2... loads:
```
|      0 | C[0][0]  | C[1][0]  | C[2][0]  | C[3][0]  | C[8][0]   | C[9][0]   | C[10][0]  | C[11][0]  | C[16][0]  | C[17][0]  | C[18][0]  | C[19][0]  | C[24][0]  | C[25][0]  | C[26][0]  | C[27][0]  |
+--------+----------+----------+----------+----------+-----------+-----------+-----------+-----------+-----------+-----------+-----------+-----------+-----------+-----------+-----------+-----------+
|      1 | C[0][1]  | C[1][1]  | C[2][1]  | C[3][1]  | C[8][1]   | C[9][1]   | C[10][1]  | C[11][1]  | C[16][1]  | C[17][1]  | C[18][1]  | C[19][1]  | C[24][1]  | C[25][1]  | C[26][1]  | C[27][1]  |
+--------+----------+----------+----------+----------+-----------+-----------+-----------+-----------+-----------+-----------+-----------+-----------+-----------+-----------+-----------+-----------+
|      2 | C[0][2]  | C[1][2]  | C[2][2]  | C[3][2]  | C[8][2]   | C[9][2]   | C[10][2]  | C[11][2]  | C[16][2]  | C[17][2]  | C[18][2]  | C[19][2]  | C[24][2]  | C[25][2]  | C[26][2]  | C[27][2]  |
```
Lane 32, 33, 34... loads:
```
|     32 | C[4][0]  | C[5][0]  | C[6][0]  | C[7][0]  | C[12][0]  | C[13][0]  | C[14][0]  | C[15][0]  | C[20][0]  | C[21][0]  | C[22][0]  | C[23][0]  | C[28][0]  | C[29][0]  | C[30][0]  | C[31][0]  |
+--------+----------+----------+----------+----------+-----------+-----------+-----------+-----------+-----------+-----------+-----------+-----------+-----------+-----------+-----------+-----------+
|     33 | C[4][1]  | C[5][1]  | C[6][1]  | C[7][1]  | C[12][1]  | C[13][1]  | C[14][1]  | C[15][1]  | C[20][1]  | C[21][1]  | C[22][1]  | C[23][1]  | C[28][1]  | C[29][1]  | C[30][1]  | C[31][1]  |
+--------+----------+----------+----------+----------+-----------+-----------+-----------+-----------+-----------+-----------+-----------+-----------+-----------+-----------+-----------+-----------+
|     34 | C[4][2]  | C[5][2]  | C[6][2]  | C[7][2]  | C[12][2]  | C[13][2]  | C[14][2]  | C[15][2]  | C[20][2]  | C[21][2]  | C[22][2]  | C[23][2]  | C[28][2]  | C[29][2]  | C[30][2]  | C[31][2]  |
```
The slice that a thread loads here is actually a slice of a `4x2x4x32` tensor (which came from a `32x32` block).

Reply:
Interesting, I did not think that much and I was exercising on the `F32_16x16x4_F32` case.  Are there are other cases that appears in A/B matrix? This makes me feel that C matrix is special. If only C is special, we can have separate `GPUSetEncodingOpLoweringConversion` pattern to handle C matrix.

I feel that it can be a follow-up. The plan in my mind is building everything around `F32_16x16x4_F32`; refactor and extend the support for other intrinsics. I feel that we will break it into two patterns (one for A/B matrix and one for C matrix) when we are adding the support for the `f32_32x32x8_f16` case.

(I can't flesh out the formula at this moment. @lialan @pashu123 please keep it in your mind or follow-up with Max later.)

Reply:
I am not sure if this appears in the A/B matrices for any intrinsics, but I think it makes sense to make the expansion generalized to different sized reshapes.

In my mind, I see the `MaterializeEncodingInfo` for GPU data tiling looking something like this:
```
struct MaterializeEncodingInfo {
  // Everything from before for the first level packing
  SmallVector<int64_t> innerDimsPos;
  SmallVector<int64_t> innerTileSizes;
  SmallVector<int64_t> outerDimsPerm;
  unsigned srcRank = 0;
  // Metadata for a generalized expand_shape + transpose
  SmallVector<SmallVector<int64_t>> innerTileShapes;
  SmallVector<int64_t> permutation;
};
```
`innerTileShapes` will hold the expanded shape for each of the inner_tile dimensions. So, the product of each `SmallVector<int64_t>` in `innerTileShapes` will be equal to the corresponding tile size in `innerTileSizes`. Then, the `permutation` tells how to swizzle those inner dimensions to what the kernel expects. From Benoit's explanations, I think this is the level of control that the kernel designer needs to have.

With this `MaterializeEncodingInfo`, the materialization pattern becomes:
- Create tensor.pack, similar to on CPU
- Expand just the inner_tile dimensions using `innerTileShapes`
- Transpose just the expanded inner_tile dims using `permutation`

We can do this because the expectation is that a kernel designer will write a kernel just for the inner tile created by tensor.pack. The `innerTileShapes` + `permutation` allows for a fully general expand_shape + transpose of the inner tile, and it maps pretty well to the expand_shape + transpose metadata while still being relatively intuitive about what swizzle is happening.

Unset_encoding will be the inverse of this:
- Create inverse transpose
- Collapse with the same reassociations, collapsing the inner shape into the `innerTileSizes`
- Create tensor.unpack, similar to on CPU

What do you guys think?

Reply:
I see, I think this is okay for now then. It should be easy to clean it up later.

Reply:
updated